library/Zend/ServiceManager/ServiceManager.php

```diff
@@ -1002,10 +1011,33 @@ protected function retrieveFromPeeringManager($name)
             } while ($this->hasAlias($name));
         }
 
+        if (null !== ($service = $this->loopPeeringServiceManagers($name))) {
+            return $service;
+        }
+
+        return null;
+    }
+
+    /**
+     * Loop over peering service managers.
+     *
+     * @param string $name
+     * @return mixed
+     */
+    protected function loopPeeringServiceManagers($name)
+    {
+        $caller = $this->serviceManagerCaller;
+
         foreach ($this->peeringServiceManagers as $peeringServiceManager) {
-            if ($peeringServiceManager->has($name)) {
-                $this->shared[$name] = $peeringServiceManager->isShared($name);
-                return $peeringServiceManager->get($name);
+            // ignore peering service manager if they are the same instance
+            if ($caller !== $peeringServiceManager) {
+                // pass this instance to peering service manager
+                $peeringServiceManager->serviceManagerCaller = $this;
+                if ($peeringServiceManager->has($name)) {
+                    $this->shared[$name] = $peeringServiceManager->isShared($name);
+                    return $peeringServiceManager->get($name);
```

Comment:
Shouldn't `$peeringServiceManager->serviceManagerCaller` be reset to `null` before returning?